---
title: "An example of predicting cell cycle phase using peco"
author: "Joyce Hsiao"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

`peco` is a supervised approach for PrEdicting cell cycle phase in a
COntinuum using single-cell RNA sequencing data in human induced
pluripotent stem cells (iPSCs). This vignette provides a simple
example of predicting cell cycle phase using 5 genes. These 5 genes
were identified as exhibiting strong cyclic trend in our analysis, and
also as cell-cycle marker genes in previous cell-cycle studies using
yeast ([Spellman et al., 1998][spellman]) and HeLa cells
([Whitfield et al., 2002][whitfield]).

## Preparation

### 1. Load data and packages

```{r, message=FALSE}
library(Biobase)
library(peco)

# expression set of top 100 cyclic genes
eset <- get(data(eset_final_sub))

# All 11,040 genes included in our dataset
# ordered by cyclic trend (strong to weak)
genes_cyclic_list <- get(data(genes_cyclic_list))
```

### 2. Data preprocessing

Note that we mapped our data to GRCh37 Ensembl Assembly. 

```{r}
# phenotype information
pdata <- pData(eset)

# gene information
fdata <- fData(eset)

# select endogeneous genes
counts <- exprs(eset)[grep("ENSG", rownames(eset)), ]

# normalize molecule count for differencese in library sizes
log2cpm <- t(log2(1+(10^6)*(t(counts)/pdata$molecules)))

# reordering the data according to FUCCI phase
log2cpm <- log2cpm[,order(pdata$theta)]
pdata <- pdata[order(pdata$theta),]


# quantile-transform each gene to normal distribution
log2cpm.quant <- do.call(rbind, lapply(1:nrow(log2cpm), function(g) {
  yy <- log2cpm[g,]
  is.zero <- which(yy == 0)
  qq.map <- qqnorm(yy, plot.it = F)
  yy.qq <- qq.map$x
  yy.qq[is.zero] <- sample(qq.map$x[is.zero])
  return(yy.qq)
}) )
rownames(log2cpm.quant) <- rownames(log2cpm)
colnames(log2cpm.quant) <- colnames(log2cpm)


# Select single-cell samples for our prediction example
which_samples_predict <- rownames(pdata)[which(pdata$chip_id == "NA19098")]
which_samples_train <- rownames(pdata)[which(pdata$chip_id != "NA19098")]

# use top 5 cyclic genes for prediciton
ngenes <- 5
which_genes <- genes_cyclic_list$ensg[1:ngenes]
```

We quantile-normalize log2CPM values for each gene to a normal
distribution. The samples with zero molecule count are each assigned a
different quantile value. By doing this, the transformed gene
expression values centered at 0 with standard deviation of 1.

```{r, fig.width=9, fig.height=7}
par(mfrow=c(2,2), mar=c(4,4,3,1))
plot(log2cpm[1,], pch=16, cex=.7,
     main="Before quantile-normalization, log2CPM values",
     xlab="FUCCI phase", ylab="Gene expression levels")
plot(log2cpm.quant[1,], pch=16, cex=.7,
     main="After quantile-normalization, log2CPM values",
     xlab="FUCCI phase", ylab="Gene expression levels")
plot(x=log2cpm[1,], y=log2cpm.quant[1,], pch=16, cex=.7,
     main="Expression values before vs after quantile-normalization",
     xlab="log2CPM before quantile-normalization", 
     ylab="log2CPM after quantile-normalization")
title(rownames(log2cpm)[1], outer=TRUE, line=-1)
```

## Analysis

### 1. Load pre-computed results

```{r}
data(fit)
```

### 2. Training

Estimating cyclic functions of the desired genes using the traing
data. In this example, we include single-cell samples from 5
individuals.

```{r}
# learning cyclic functions of the genes using our training data
Y_train <- log2cpm.quant[which(rownames(log2cpm.quant) %in% which_genes), 
                         which(colnames(log2cpm.quant) %in% which_samples_train)]
theta_train <- pdata$theta[which(rownames(pdata) %in% which_samples_train)]
names(theta_train) <- rownames(pdata)[which(rownames(pdata) %in% which_samples_train)]
```

```{r, eval=FALSE}
fit_train <- cycle_npreg_insample(Y = Y_train,
                                  theta = theta_train,
                                  polyorder=2,
                                  ncores=1,
                                  method.trend="trendfilter")
```

Fitted cyclic trend in the training data.

```{r, fig.width=9, fig.height=8}
par(mfrow=c(2,3), mar=c(4,4,3,1))
for (g in 1:5) {
plot(fit_train$Y[g,], 
     x=fit_train$theta, axes=F,
     ylab="Gene expression values (log2CPM quantile-normalized)")
points(y=fit_train$funs_est[[g]](fit_train$theta),
       x=fit_train$theta,
       pch=16, col="royalblue")
axis(2); axis(1,at=c(0,pi/2, pi, 3*pi/2, 2*pi),
              labels=c(0,expression(pi/2), expression(pi), expression(3*pi/2),
                       expression(2*pi)))
abline(h=0, lty=1, col="black", lwd=.7)
title(rownames(fit_train$Y)[g])
}
title("Cyclic trends in the training data", outer=T, line=-1)
```

### 3. Predicting

Predicting cell cycle phase for single-cell samples from one individual.

```{r}
# predicting cell cycle phase for single-cell samples from one individual
# that is not included in the training 
Y_predict <- log2cpm.quant[which(rownames(log2cpm.quant) %in% which_genes),
                           which(colnames(log2cpm.quant) %in% which_samples_predict)]

theta_test <- pdata$theta[which(rownames(pdata) %in% which_samples_predict)]
names(theta_test) <- rownames(pdata)[which(rownames(pdata) %in% which_samples_predict)]
```

```{r, eval=FALSE}
fit_predict <- cycle_npreg_outsample(Y_test=Y_predict,
                                  sigma_est=fit_train$sigma_est,
                                  funs_est=fit_train$funs_est,
                                  theta_prior=fit_train$theta,
                                  method.grid = "uniform",
                                  method.trend="trendfilter",
                                  polyorder=2, ncores=1)
```


Plot predicted phase versus FUCCI phase 

```{r, fig.width=9, fig.height=8}
theta_test_reordered <- theta_test[match(names(fit_predict$cell_times_reordered),
                                         names(theta_test))]
all.equal(names(fit_predict$cell_times_reordered), names(theta_test_reordered))

par(mfrow=c(2,3), mar=c(4,4,3,1))
for (g in 1:5) {
plot(fit_predict$Y_reordered[g,], 
     x=fit_predict$cell_times_reordered, axes=F,
     xlab="FUCCI phase",
     ylab="Predicted phase")
points(y=fit_predict$funs_reordered[[g]](fit_predict$cell_times_reordered), 
       x=fit_predict$cell_times_reordered,
       pch=16, col="royalblue")
axis(2); axis(1,at=c(0,pi/2, pi, 3*pi/2, 2*pi),
              labels=c(0,expression(pi/2), expression(pi), expression(3*pi/2),
                       expression(2*pi)))
abline(h=0, lty=1, col="black", lwd=.7)
title(rownames(fit_predict$Y_reordered)[g])
}
title("Predicting cell cycle phase for NA19160", outer=T)
```

## Session information

```{r}
sessionInfo()
```

[spellman]: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC25624
[whitfield]: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC117619
