---
title: "Predicting cell cycle phase using peco"
author: "Joyce Hsiao"
date: "`r Sys.Date()`"
output: 
    rmarkdown::html_vignette
vignette: >
    %\VignetteIndexEntry{An example of predicting cell cycle phase using peco}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```

`peco` is a supervised approach for PrEdicting cell cycle phase in a
COntinuum using single-cell RNA sequencing data. This vignette provides a 
simple example of predicting cell cycle phase using 5 cyclic genes. We 
identifeid the 5 genes as the top 5 cyclic genes in our study, which also 
showed thata simple predictor of top 5 cyclic genes performed as well as when 
including more genes with weaker cyclic trends. 

These five genes are: _CDK1_, _UBE2C_, _TOP2A_, _HISTH1E_, and _HISTH1C_, 
which were identified as cell cycle marker genes in previous studies of 
the cell cycle using yeast ([Spellman et al., 1998][spellman]) and 
HeLa cells ([Whitfield et al., 2002][whitfield]).


##  Load data and packages

```{r, message=FALSE}
library(SingleCellExperiment)
library(peco)
```

## Predicting cell cycle using pre-trained data

`training_human` contains pre-trained results from 888 single-cell sample,
including gene-specific cyclic functions.

```{r}
data("training_human")
str("training_human")
```

Below shows an example of inputting `SingleCellExperiment` object to perform
cell cycle phase prediction. For now, `peco` can be immediatley applied for
expression data that stores Ensembl ID as default gene labels. In this example,
`sce_top101genes` is the input data.

First transform the expression values to quantile-normalizesd CPM values.

```{r}
data("sce_top101genes")
sce_top101genes <- data_transform_quantile(sce_top101genes)
```

Apply the prediction model.

```{r}
model_101genes_predict <- cycle_npreg_outsample(
    Y_test=sce_top101genes,
    sigma_est=training_human$sigma[rownames(sce_top101genes),],
    funs_est=training_human$cellcycle_function[rownames(sce_top101genes)],
    method.trend="trendfilter",
    ncores=1,
    get_trend_estimates=FALSE)
```

`cycle_npreg_outsample` outputs a SingleCellExpression object that contains
predicted cell cycle phase.

```{r}
colData(model_101genes_predict)$cellcycle_peco
```

Visualize results of prediction for one gene. Below we choose CDK1. Because
CDK1 is a known cell cycle gene, this visualization serves as a sanity
check for the results of fitting. 

```{r}
plot(y=assay(sce_top101genes,"cpm_quantNormed")[1,],
     x=colData(sce_top101genes)$theta_shifted, main = "CDK1")
points(y=training_human$cellcycle_function[[1]](seq(0,2*pi, length.out=100)),
       x=seq(0,2*pi, length.out=100), col = "blue", pch =16)
```


## Session information

```{r}
sessionInfo()
```

[spellman]: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC25624
[whitfield]: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC117619
