---
title: "An example of predicting cell cycle phase using peco"
author: "Joyce Hsiao"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{An example of predicting cell cycle phase using peco}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

`peco` is a supervised approach for PrEdicting cell cycle phase in a
COntinuum using single-cell RNA sequencing data. This vignette provides a 
simple example of predicting cell cycle phase using 5 cyclic genes. We 
identifeid the 5 genes as the top 5 cyclic genes in our study, which also 
showed thata simple predictor of top 5 cyclic genes performed as well as when 
including more genes with weaker cyclic trends. 

These five genes are: _CDK1_, _UBE2C_, _TOP2A_, _HISTH1E_, and _HISTH1C_, 
which were identified as cell cycle marker genes in previous studies of 
the cell cycle using yeast ([Spellman et al., 1998][spellman]) and 
HeLa cells ([Whitfield et al., 2002][whitfield]).


## Preparation

### 1. Load data and packages

```{r, message=FALSE}
library(SingleCellExperiment)
library(peco)

# expression set of top 101 cyclic genes
data(sce_top101genes)
```

### 2. Data preprocessing

Note that we mapped our data to GRCh37 Ensembl Assembly. 

```{r}
# phenotype information
coldata <- colData(sce_top101genes)

# gene information
rowdata <- rowData(sce_top101genes)

# select top 5 cyclic genes
sce_top5 <- sce_top101genes[order(rowData(sce_top101genes)$pve_fucci,
                                  decreasing=TRUE)[1:5],]

coldata <- colData(sce_top5)[order(colData(sce_top5)$theta_shifted),]

# quantile-transform CPM to normal distribution
sce_quant <- data_transform_quantile(sce_top5)
expr_quant <- assay(sce_quant, "cpm_quant")
counts_normed <- assay(sce_quant, "cpm")
```

We quantile-normalize CPM values for each gene to a normal
distribution. The samples with zero molecule count are each assigned a
different quantile value. By doing this, the transformed gene
expression values centered at 0 with standard deviation of 1.

```{r, fig.width=9, fig.height=7}
par(mfrow=c(2,2), mar=c(4,5,4,1))
plot(counts_normed[1,], pch=16, cex=.7,
     main="Before quantile-normalization",
     xlab="FUCCI phase", ylab="CPM values")
plot(expr_quant[1,], pch=16, cex=.7,
     main="After quantile-normalization",
     xlab="FUCCI phase", ylab="Quantile-normalized \n expression values")
plot(x=counts_normed[1,], y=expr_quant[1,], pch=16, cex=.7,
     main="Before vs after quantile-normalization",
     xlab="CPM values", 
     ylab="Quantile-normalized \n expression values")
title(rownames(counts_normed)[1], outer=TRUE, line=-1)
```

## Analysis

### 1. Load pre-computed results

```{r}
data("model_5genes_train")
data("model_5genes_predict")
```

### 2. Training

Estimating cyclic functions of the desired genes using the traing
data. In this example, we include single-cell samples from 5
individuals.

```{r, eval=FALSE}
# Select samples from NA18511 for our prediction example
which_samples_train <- rownames(coldata)[coldata$chip_id != "NA18511"]
which_samples_predict <- rownames(coldata)[coldata$chip_id == "NA18511"]

# learning cyclic functions of the genes using our training data
Y_train <- expr_quant[, colnames(expr_quant) %in% which_samples_train]
theta_train <- 
    coldata$theta_shifted[rownames(coldata) %in% which_samples_train]
names(theta_train) <- 
    rownames(coldata)[rownames(coldata) %in% which_samples_train]

model_5genes_train <- cycle_npreg_insample(Y = Y_train,
                                  theta = theta_train,
                                  polyorder=2,
                                  ncores=1,
                                  method.trend="trendfilter")
```

Fitted cyclic trend in the training data.

```{r, fig.width=9, fig.height=8}
par(mfrow=c(2,3), mar=c(4,4,3,1))
for (g in 1:5) {
plot(model_5genes_train$Y[g,], 
     x=model_5genes_train$theta, axes=FALSE,
     ylab="quantile-normalized expressio values")
points(y=model_5genes_train$funs_est[[g]](model_5genes_train$theta),
       x=model_5genes_train$theta,
       pch=16, col="royalblue")
axis(2); axis(1,at=c(0,pi/2, pi, 3*pi/2, 2*pi),
              labels=c(0,expression(pi/2), expression(pi), expression(3*pi/2),
                       expression(2*pi)))
abline(h=0, lty=1, col="black", lwd=.7)
title(rownames(model_5genes_train$Y)[g])
}
title("Cyclic trends in the training data", outer=TRUE, line=-1)
```

### 3. Predicting

Predicting cell cycle phase for single-cell samples from one individual.

```{r, eval=F}
# predicting cell cycle phase for single-cell samples from one individual
# that is not included in the training 
##------ Predict cell cycle phase
Y_predict <- expr_quant[, colnames(expr_quant) %in% which_samples_predict]
model_5genes_predict <- cycle_npreg_outsample(Y_test=Y_predict,
                                     sigma_est=model_5genes_train$sigma_est,
                                     funs_est=model_5genes_train$funs_est,
                                     method.trend="trendfilter",
                                     ncores=1,
                                     get_trend_estimates=TRUE)
```

Results of prediction in the testing sample. 

```{r, fig.width=9, fig.height=8}
par(mfrow=c(2,3), mar=c(4,4,3,1))
for (g in 1:5) {
plot(model_5genes_predict$Y_reordered[g,], 
     x=model_5genes_predict$cell_times_reordered, axes=FALSE,
     xlab="FUCCI phase",
     ylab="Predicted phase")
points(
  y=model_5genes_predict$funs_reordered[[g]](
        model_5genes_predict$cell_times_reordered),
  x=model_5genes_predict$cell_times_reordered,
     pch=16, col="royalblue")
axis(2); axis(1,at=c(0,pi/2, pi, 3*pi/2, 2*pi),
              labels=c(0,expression(pi/2), expression(pi), expression(3*pi/2),
                       expression(2*pi)))
abline(h=0, lty=1, col="black", lwd=.7)
title(rownames(model_5genes_predict$Y_reordered)[g])
}
title("Predicting cell cycle phase for NA18511", outer=TRUE)
```

## Session information

```{r}
sessionInfo()
```

[spellman]: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC25624
[whitfield]: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC117619
